Acorn Tracking

UvuvDev opened the pull request:
#### Summary
<!-- Provide a brief summary on the pull request -->

Added Acorn Tracking, derived from movement class. It takes in a vision sensor and a vision sensor signature as parameters. Gets global coordinates of acorns 

#### Motivation
<!-- Why are you making this pull request? -->

ion know

#### Test Plan
<!-- How did you test / plan to test this pull request? -->

made an app and tested it with opencv and my camera

#### Additional Notes
<!-- Add any other information you think is relevant -->

none

<!-- DO NOT REMOVE!! -->
<!-- bot: nightly-link -->
<!-- commit-sha: 910f9e0449c91e78a8e17a1eb67c71962843caa5 -->
## Download the template for this pull request: 

> [!NOTE]  
> This is auto generated from [`Add Template to Pull Request`](https://github.com/LemLib/LemLib/actions/runs/7510387748)
- via manual download: [LemLib@0.5.0-rc2+910f9e.zip](https://nightly.link/LemLib/LemLib/actions/artifacts/1166935943.zip)
- via PROS Integrated Terminal: 
 ```
curl -o LemLib@0.5.0-rc2+910f9e.zip https://nightly.link/LemLib/LemLib/actions/artifacts/1166935943.zip;
pros c fetch LemLib@0.5.0-rc2+910f9e.zip;
pros c apply LemLib@0.5.0-rc2+910f9e;
rm LemLib@0.5.0-rc2+910f9e.zip;
```

UvuvDev committed:
Acorn Tracking Implemented

Acorn Tracking is ported from uvuvlib. Derived from movement, it gets you the coordinates of the leftmost acorn seen.

liam-teale committed:
Merge branch 'refactor' into refactor

liam-teale committed:
Merge branch 'refactor' into refactor

UvuvDev committed:
New Controller Wrapper, LEMController

LEMController has the functionality of a PROS controller, but also supports shift keys, toggles, new button presses, and automatically running functions on button presses

UvuvDev committed:
Merge branch 'refactor' of https://github.com/DancingRicardo/LemLib into refactor

UvuvDev committed:
Did some edits to controller, Added Reader Auton

Reader Auton, can use Lem Drivetrains or user's custom functions to move the drive based off previous inputs. Custom functions now require an int parameter, if user doesn't want to use it it can just be left blank. just adds more options.

UvuvDev committed:
Unit Tests for controller

added unit tests for controller functions.

UvuvDev committed:
AccuracySIM psuedocode

currently non functional, written in psuedocode. Class is in place, will work on inheriting from lemlib movement class soon.  Have to port the Poller from UvuvLib as well.

UvuvDev committed:
Bug Fixes, LEMControllerValues Struct

Fixed bugs, mainly bugs stemming from not saving button states to memory. This meant that functions could keep track of one button instead of 12 like its supposed to.

This was solved by introducing a hashmap of button types and booleans, which kept track of each individual buttons state. LEMControllerValues was uncommented and reformatted to grab the keys based off a button parameter, meaning the code will now be much cleaner.

UvuvDev committed:
Bug Fix, Unit test fixes

Previously, controller was unable to run functions when buttons were false, which would mean you lose half the functionality. Fixed by making every function pointer in the code a pair now, first being for when the function is false, the second being for when it is true.

This bug fix meant some of the unit tests needed to be redone, which they are now.

UvuvDev committed:
Added function for getting SD Card data

UvuvDev committed:
Accuracy Sim Progress, needs to be tested

Added most functionality into accuracy sim. Only thing left out is integration with Jerry.IO to get the path and path points.

UvuvDev committed:
Added Jerry.IO integration, Fully Functional accuracy sim?

Took pure pursuit loading code and placed it in the constructor. Accuracy sim should be fully functional assuming no bugs.

UvuvDev committed:
Input Reader Bug Fixes, LEM Controller Bug Fixes

Due to the LEM Controller getting bug fixes yesterday, some functions were changed and the input reader was rendered non functional. LEM Controller had some more bugs, relating to the vector of LEM Button Mappings. One bug was fixed, there is still one more.

UvuvDev committed:
Controller Bug Fixes, ShapesNCoords source code, Sucessful Compiling

Finished fixing the constructor errors from last night. I added source code for the shapes currently being used by LemLib, changed some of the classes slightly, and now the code compiles. Ready to test!

UvuvDev committed:
Added Joystick Support

Joysticks were for some reason unsupported, so I added functionality for it.

UvuvDev committed:
LEMController (now gamepad) debugged, tested (works)

Renamed the LEMController to gamepad due to popular consensus. Tested functionality, including modes, auto running functions, and inputting functions. all work.

meisZWFLZ commented on src/lemlib/movements/acorntracking.cpp:
```diff
@@ -0,0 +1,60 @@
+
+        float distance = y1 + ((trackedAcornRadius-x1) * (y2-y1))/ (x2-x1);
+        // Linear Interpolation: y1 + [(x-x1) * (y2-y1)]/ (x2-x1)
+        std::pair<float, float> newcoords = std::make_pair<float, float>(pose.x + distance*std::cos(pose.theta), pose.y + distance*std::sin(pose.theta));
```
The output of the movement::update() is intended to be the power of the left and right sides of the chassis..

UvuvDev committed:
Pid Tuner

i had already pushed this, but ill push it again...

UvuvDev committed:
Incorrect Class Name Bug Fixes

Fixed some bugs where LEMController was still hanging around in the code.

UvuvDev committed:
Ready to merge

Seperated all other new features from acorn tracking.

UvuvDev committed:
Input Reader

UvuvDev committed:
fixed files

wrong files, oops

UvuvDev committed:
format

UvuvDev committed:
Took out extra files

UvuvDev committed:
format

UvuvDev committed:
Added acorn tracking functions, debugging for acorn tracking

Movement functions weren't being inherited, so class couldnt be initialized. fixed that. tried debugging acorn tracking. current issue is that the vision sensor doesnt seem to register objects, causing a NAN error.
